lib/location/method_gnss.c

```diff
@@ -0,0 +1,835 @@
+		k_work_submit_to_queue(loc_core_work_queue_get(), &method_gnss_notify_pgps_work);
+#endif
+	}
+#endif // defined(CONFIG_LOCATION_METHOD_GNSS_AGPS_EXTERNAL)
+}
+#endif // defined(CONFIG_NRF_CLOUD_AGPS) || defined(CONFIG_NRF_CLOUD_PGPS)
+
+void method_gnss_event_handler(int event)
+{
+	switch (event) {
+	case NRF_MODEM_GNSS_EVT_PVT:
+		k_work_submit_to_queue(loc_core_work_queue_get(), &method_gnss_pvt_work);
+		break;
+
+	case NRF_MODEM_GNSS_EVT_AGPS_REQ:
+#if defined(CONFIG_NRF_CLOUD_AGPS) || defined(CONFIG_NRF_CLOUD_PGPS)
+		method_gnss_request_assistance();
+#endif
+		break;
+	}
+}
+
+int method_gnss_cancel(void)
+{
+	int err = nrf_modem_gnss_stop();
+	int sleeping;
+	int rrc_idling;
+
+	if (err == NRF_EPERM) {
+		/* Let's convert NRF_EPERM to EPERM to make sure loc_core works properly */
+		err = EPERM;
+		/* NRF_EPERM is logged in loc_core */
+	} else if (err) {
+		LOG_ERR("Failed to stop GNSS");
+	}
+
+	running = false;
+
+	/* Cancel any work that has not been started yet */
+	(void)k_work_cancel(&method_gnss_start_work.work_item);
+
+	/* If we are currently not in PSM, i.e., LTE is running, reset the semaphore to unblock
+	 * method_gnss_positioning_work_fn() and allow the ongoing location request to terminate.
+	 * Otherwise, don't reset the semaphore in order not to lose information about the current
+	 * sleep state.
+	 */
+	sleeping = k_sem_count_get(&entered_psm_mode);
+	if (!sleeping) {
+		k_sem_reset(&entered_psm_mode);
+	}
+
+	/* If we are currently in RRC connected mode, reset the semaphore to unblock
+	 * method_gnss_positioning_work_fn() and allow the ongoing location request to terminate.
+	 * Otherwise, don't reset the semaphore in order not to lose information about the current
+	 * RRC state.
+	 */
+	rrc_idling = k_sem_count_get(&entered_rrc_idle);
+	if (!rrc_idling) {
+		k_sem_reset(&entered_rrc_idle);
+	}
+
+
+	return -err;
+}
+
+#if !defined(CONFIG_NRF_CLOUD_AGPS)
+static bool method_gnss_entered_psm(void)
+{
+	int ret = 0;
+	int tau;
+	int active_time;
+	enum lte_lc_system_mode mode;
+
+	lte_lc_system_mode_get(&mode, NULL);
+
+	/* Don't care about PSM if we are in GNSS only mode */
+	if (mode != LTE_LC_SYSTEM_MODE_GPS) {
+		ret = lte_lc_psm_get(&tau, &active_time);
+		if (ret < 0) {
+			LOG_ERR("Cannot get PSM config: %d. Starting GNSS right away.", ret);
+			return true;
+		}
+
+		LOG_DBG("LTE active time: %d seconds", active_time);
+
+		if (active_time >= 0) {
+			LOG_DBG("Waiting for LTE to enter PSM...");
+
+			/* Wait for the PSM to start. If semaphore is reset during the waiting
+			 * period, the position request was canceled.
+			 */
+			if (k_sem_take(&entered_psm_mode, K_MINUTES(SLEEP_WAIT_BACKSTOP))
+			    == -EAGAIN) {
+				if (!running) { /* Location request was cancelled */
+					return false;
+				}
+				/* We're still running, i.e., the wait for PSM timed out */
+				LOG_ERR("PSM is configured, but modem did not enter PSM ");
+				LOG_ERR("in %d minutes. Starting GNSS anyway.",
```

Comment:
Why is this divided into two separate log lines? And I guess a LOG_WRN might be more appropriate.